### ISSUE TITLE
Directory Structure

### ISSUE THREAD

**Issue by lirantal** [lirantal/Riess.js#3]
In order to de-couple the framework business form the app/business logic business I was thinking of  creating the following directory structure:

```
 - lib/    - all the library bootstrapping, services, etc, goes here
 - lib/core/express.js
 - lib/core/config.js
 - lib/connectors/mongoose.js
 - app/    - user own app logic. inside there will be a modules/ directory to hold a vertical modules structure
 - app/modules/
 - app/modules/helloworld/
 - app/modules/helloworld/controllers/
 - app/modules/helloworld/routes/
 - app/modules/helloworld/tests/
 - db/     - database migrations, seeds, etc. will probably need to support more than one db (sequelize, mongoose, etc)
 - db/mongodb/migrations/
 - db/sql/migrations/
```

**Comment by lirantal** [comment]
@simison @mleanos would love to get some input here

**Comment by simison** [comment]
I like that.

Thoughts yet on place for configs? In current mean.js I dislike how configs folder has lots of files with logic in them.

**Comment by lirantal** [comment]
I have some stuff to say on config management so I'll open a new thread on that.

How would you place framework files under `lib/`? All the `config/lib/*` and configuration management? Would you just place all of them into `lib/handlers/` maybe?

**Comment by simison** [comment]
I'm imagining something like this:
```
lib
├── app.js
├── worker.js
├── config.js
├── data
│   └── languages.json  — probably better under config?
└── services
    ├── cron.js
    ├── express.js
    ├── facebook-api.js
    ├── firebase-api.js
    ├── logger.js
    ├── mongoose.js
    ├── multer.js
    ├── socket.js
    └── render.js
```

I'm just adding a bunch files we have at Trustroots so it's easier to imagine how people might expand this.

I kinda feel like putting all those initializers from current meanjs `express.js` onto their own files: `express.session.js` or `express/session.js`?

**Comment by simison** [comment]
I updated above, I made a few mistakes there.

And `config.js` should probably be here too, instead of `/config/index.js`.

**Comment by lirantal** [comment]
Sounds good, but I wonder if express app initialization should happen last so it's not "purely" a service but rather part of the core framework.

For example, I can imagine mongodb, and redis being services (or connectors), and then all services are instantiated first which would translate to connecting to mongodb, loading all models, etc, and only when all those services are up then we run initialize the express app.

What lead me to do that is 2 things:
1. express can be dependent on those services, for example requiring a mongodb configuration/instance for the session store.
2. when you want to test express with supertest() then you want to really just get an app of express without the boostrapping process that gives you the app to require connecting to mongodb, redis, etc.

Ending up with something along the lines of:
```
server.startServices() // mongodb, redis, logger, etc. all can register themselves as singletons in some services object
server.startApp() // starts the expressjs app. can access the services object to get mongodb like services.get('mongodb') an gets the service by name.
```

**Comment by simison** [comment]
> I wonder if express app initialization should happen last so it's not "purely" a service but rather part of the core framework.

>express can be dependent on those services, for example requiring a mongodb configuration/instance for the session store.

Good question. For testing you want everything to be atomic — imagine running all the rests related to all those services and all the Express stuff (sessions, Helmet, etc) independently in parallel.

Ideally there should not be need to bootstrap the whole app for each test file.

Check https://github.com/avajs/ava

**Comment by lirantal** [comment]
yeah I work with ava.js at work and for a few other open source projects, but I wasn't sure if you were pointing out a different way of approaching it or not :-)
you're welcome to reference another implementation of bootstrapping expressjs which makes sense from testing in parallel.

p.s. I'll open a new thread on testing since you brought it up :-)

**Comment by lirantal** [comment]
@simison might want to take a look at some cleanup and refactoring I've done in recent commits
